docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,211 @@
+   determines if an upgrade is required. If that's the case, the LNS will
+   download the upgrade file from storage and send a properly populated response
+   to the Basics Station.
+
+1. The Basics Station will then execute the actual firmware upgrade.
+
+```mermaid
+  sequenceDiagram
+    autonumber
+    User->>LoRa Device Provisioning CLI: Request Station upgrade
+    LoRa Device Provisioning CLI->>Storage account: Upload firmware file
+    LoRa Device Provisioning CLI->>IoT Hub: Update 'Concentrator' twin
+    Concentrator->>Network Server: POST /update-info
+    Network Server->>IoT Hub: Retrieve 'Concentrator' twin
+    Network Server->>Network Server: Check if versions of the Station are different
+    alt versions are the same
+        Network Server->>Concentrator: CUPS response without firmware upgrade
+    else versions are different
+      Network Server->>Network Server: Verify if any of the `keys` from CUPS request match the `fwKeyChecksum`
+      alt no match
+        Network Server->>Network Server: Throw as this indicates a misconfiguration
+      else one of the keys matches `fwKeyChecksum`
+        Network Server->>Facade Function: Retrieve firmware upgrade file
+        Facade Function->>Storage account: Download firmware file
+        Network Server->>Concentrator: CUPS response with firmware upgrade fields populated
+      end
+    end
+```
+
+### IoT Hub related changes
+
+Only change is related to the concentrator device twin.
+
+In addition to the values already stored in the twin, the following need to be
+added:
+
+```json
+"cups": {
+  // ...
+  "package": "1.0.1",
+  "fwUrl": "https://...",
+  "fwKeyChecksum": 123456,
+  "fwSignature": "..."
+}
+```
+
+- **'package'**: desired package version of the Station (matching what will be
+  extracted as version.txt file during update)
+- **'fwUrl'**: URL pointing to the storage location of the file required to run
+  the upgrade
+- **'fwKeyChecksum'**: checksum of the key used to sign the digest of the
+  firmware upgrade file
+- **'fwSignature'**: signature of the uploaded firmware upgrade file (as base64
+  encoded string)
+
+### Storage related changes
+
+We will use the same storage solution as the one selected for the general CUPS
+protocol support (storage account). A new container named `"fwupgrades"` will be
+used to store the firmware upgrade files.  The file names will be in the format
+`"{stationEui}-{package}"` (without any extension, as anyways these are going to
+be downloaded as `update.bin` from the Basics Station executable).
+
+### Azure Function related changes
+
+A new endpoint will be added in the Facade Azure Function which will be used to
+fetch firmware upgrade files from the storage account. The endpoint will accept
+the `StationEui` as input, then retrieve the concentrator twin from IoT Hub,
+download the firmware file from the storage account and send it back in the
+response.
```

Comment:
this should be a stream not a download and send. Just making sure we are on the same page. We do not want to put the storage content in memory.